### ISSUE TITLE
How to increase recall

### ISSUE THREAD

**Issue by jaqub-manuel** [ultralytics/yolov5#2449]
Dear, @glenn-jocher 
Although I have done many trials, the recall value is low compared to the precision value. Although I set the recall value of the fitness function to 0.8 and the precision value to 0.2, I did not get a good result.
Can you share some tips if you have some suggestions?
thanks.

https://github.com/ultralytics/yolov5/blob/886f1c03d839575afecb059accf74296fad395b6/utils/metrics.py#L12

![results](https://user-images.githubusercontent.com/64843726/110993724-b6c3ac80-8388-11eb-9c51-2c9e463e0452.png)

**Comment by glenn-jocher** [comment]
@jaqub-manuel sure I can share some tips! Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users first train with all default settings before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥1.5k images per class
- **Instances per class.** ≥10k instances (labeled objects) per class total
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x will produce better results in nearly all cases, but have more parameters and are slower to run. For mobile applications we recommend YOLOv5s/m, for cloud or desktop applications we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

To start training from pretrained weights simply pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
```

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://user-images.githubusercontent.com/26833433/103595982-ab986000-4eb1-11eb-8c57-4726261b0a88.png"></p>

## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by jaqub-manuel** [comment]
Dear @glenn-jocher , many thanks for your quick reply.
My dataset consists of medical data set and 700 images (416x384) for train and test set consists of 200 images (1200x900) and labeled by expert doctors.
I have trained Yolov5s for m, l and x but generally s and m works better than l and x. Everything I used default first, then I used hyp.finetune. I run each training at least 5 times for the same model and each time has different results. (different, map=i.e 70+- 10)
I did hyperparameter optimization with genetic algorithm, but still the recall value in the results is small. i am stitching all the pictures. I am interested in F1, TP, FP, FN instead of mAP. (it is medical data)
Thank you in advance for your help...




![results](https://user-images.githubusercontent.com/64843726/111044183-15a42700-8458-11eb-8f73-ab715e4955bc.png)
![confusion_matrix](https://user-images.githubusercontent.com/64843726/111044184-16d55400-8458-11eb-9801-be77354573c3.png)
![F1_curve](https://user-images.githubusercontent.com/64843726/111044185-176dea80-8458-11eb-8c89-9c25a47107ae.png)
![labels](https://user-images.githubusercontent.com/64843726/111044186-18068100-8458-11eb-86e2-8c4ce418e729.jpg)
![labels_correlogram](https://user-images.githubusercontent.com/64843726/111044187-189f1780-8458-11eb-8a55-bb573deded3c.jpg)
![P_curve](https://user-images.githubusercontent.com/64843726/111044188-1937ae00-8458-11eb-825a-3265bb6eee26.png)
![PR_curve](https://user-images.githubusercontent.com/64843726/111044190-1937ae00-8458-11eb-8d67-33821d8a6f94.png)
![R_curve](https://user-images.githubusercontent.com/64843726/111044191-19d04480-8458-11eb-8635-ce9b89cf1eb2.png)

**Comment by glenn-jocher** [comment]
@jaqub-manuel yes your recall is low. The quantity of your data is about an order of magnitude too small vs other datasets like COCO, and also your image sizes differ substantially between your train and test sets. You want your training data and your val/test data to be sampled from the same image space, and they should share similar characteristics and variations.

Also better results with smaller models is likely independently indicative of a problem in your training pipeline.

You may also want to use P6 models, which outperform their P5 counterparts across the board:
```
python train.py --weights yolov5s6.pt
```

<img width="2029" alt="Screen Shot 2021-01-07 at 9 18 07 PM" src="https://user-images.githubusercontent.com/26833433/111044755-5754b380-83ff-11eb-852b-cc4ed2cb38cc.png">

**Comment by glenn-jocher** [comment]
@jaqub-manuel also you can evolve hyperparameters with a custom fitness definition better suited to your requirements. The default fitness will favor mutations that produce higher mAP@0.5:0.95 mostly. You can shift this to F1 or mAP@0.5 to reduce the influence of accurate bounding boxes by modifying the fitness function here:
https://github.com/ultralytics/yolov5/blob/747c2653eecfb870b1ed40b1e00e0ef209b036e9/utils/metrics.py#L12-L16

Though in general you will obtain much better results via dataset improvement than training improvement considering the small size of your dataset.

**Comment by jaqub-manuel** [comment]
Many thanks @glenn-jocher, 
I will try and upload if I get better results.

**Comment by jaqub-manuel** [comment]
@glenn-jocher, I am now using evolve, 
I wonder if the below code is correct if it is only F1 oriented?

 def fitness(x): 
     # Model fitness as a weighted combination of metrics 
     **w = [0.0, 0.0, 0.0, 0.0, 1.0]**  # weights for [P, R, mAP@0.5, mAP@0.5:0.95] 
     return (x[:, :4] * w).sum(1)

**Comment by di-dii** [comment]
@jaqub-manuel  
do you get a better result by doing evolution of hyperparameters?

**Comment by jaqub-manuel** [comment]
hey @di-dii,
I always got different result because I have about 500 images for training (it should be over 1000 for better result). For fine tune, I did not get better results.

**Comment by Lg955** [comment]
Hi @glenn-jocher  , I have a question: After training, the F1_curve shows that "all classes 0.82 at 0.444", so the conf-thres should be set as 0.444 when detecting?
![F1_curve](https://user-images.githubusercontent.com/69025725/116257472-ab94d680-a7a6-11eb-9027-01f33f45a5fb.png)
![PR_curve](https://user-images.githubusercontent.com/69025725/116257492-af285d80-a7a6-11eb-8f16-a967a0269288.png)
![results](https://user-images.githubusercontent.com/69025725/116257515-b18ab780-a7a6-11eb-84bd-e486daac98ef.png)

**Comment by glenn-jocher** [comment]
@Lg955 0.444 confidence threshold produces the maximum F1 value for your dataset and is a good point for setting your deployed inference confidence threshold.

**Comment by Lg955** [comment]
> @Lg 9550.444置信阈值为数据集生成最大F1值，是设置已部署的推理置信阈值的好点。

If I want to get the highest mAP,  0.444 for conf-thres is the best?

**Comment by glenn-jocher** [comment]
@Lg955 no. The highest mAP will be produced at confidence threshold 0.0.

**Comment by Lg955** [comment]
> @Lg955 no. The highest mAP will be produced at confidence threshold 0.0.

Oh, I see.  Thank you!

**Comment by 1chimaruGin** [comment]
@glenn-jocher  Is there a tip to improve model recall if object and background are similar. In my case val/obj loss is keep increasing.

![W B Chart 1_7_2022, 9_47_17 PM](https://user-images.githubusercontent.com/35499653/148565458-58dd4443-3382-42db-8b4a-967b26636d34.png)

![b77c1347-defc-4842-a22f-6a102b82d57f](https://user-images.githubusercontent.com/35499653/148565465-2fbc5c69-b753-4441-ace1-849d1120466a.png)

![c67c0404-fcf2-4798-8711-1df26225d7df](https://user-images.githubusercontent.com/35499653/148565518-952a3311-41e2-470c-94c5-1561c1353af0.png)
![4111daa9-1a17-4bea-849d-111de8a1fbf1](https://user-images.githubusercontent.com/35499653/148565540-ec920048-c145-46f6-9722-28ad82bb3585.png)

**Comment by glenn-jocher** [comment]
@1chimaruGin 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/](https://github.com/ultralytics/yolov5/issues/2844#issuecomment-851338384)

Good luck 🍀 and let us know if you have any other questions!

**Comment by 1chimaruGin** [comment]
Thanks for the quick response

Here my training results. 

https://wandb.ai/1chimarugin/YOLOv5?workspace=user-1chimarugin

**Comment by srn-source** [comment]
> @glenn-jocher, I am now using evolve, I wonder if the below code is correct if it is only F1 oriented?
> 
> def fitness(x): # Model fitness as a weighted combination of metrics **w = [0.0, 0.0, 0.0, 0.0, 1.0]** # weights for [P, R, mAP@0.5, mAP@0.5:0.95] return (x[:, :4] * w).sum(1)

@glenn-jocher 
If i want to monitor only F1, do i have to change something like this?

**Comment by glenn-jocher** [comment]
@srn-source sure

**Comment by glenn-jocher** [comment]
IoU can be tuned as well. Default detect.py value is 0.45.
